subsys/net/lib/aws_fota/src/aws_fota.c

```diff
@@ -341,69 +429,174 @@ static int on_publish_evt(struct mqtt_client *const client,
+			goto read_payload;
+		}
 
+		if (!is_job_id_in_topic(topic, topic_len, job_id_handling)) {
+			LOG_WRN("The currently handled job ID is not in incoming rejected topic");
+			goto read_payload;
+		}
+
+		return job_update_rejected(client, payload_len);
+	default:
+		/* The incoming topic is not related to AWS FOTA. */
+		return 1;
+	}
+
+read_payload:
+
+	/* If we get here, a job is already being handled. The incoming
+	 * message is most likely a duplicate and must be dropped
+	 * to not interfere with the job being processed.
+	 * The payload must still be read out in order to clear the
+	 * MQTT buffer for next incoming message.
+	 */
+	err = get_published_payload(client, payload_buf, payload_len);
+	if (err) {
+		LOG_ERR("Error when getting the payload: %d", err);
+		return err;
+	}
+
+	LOG_DBG("FOTA already in progress, message is ignored");
+	return 0;
+}
+
+static int on_connack_evt(struct mqtt_client *const client)
+{
+	int err;
+
+	switch (internal_state) {
+	case STATE_INIT:
+		err = aws_jobs_subscribe_topic_notify_next(client, notify_next_topic);
 		if (err) {
-			LOG_ERR("Error when getting the payload: %d", err);
+			LOG_ERR("Unable to subscribe to notify-next topic");
 			return err;
 		}
 
-		LOG_DBG("FOTA already in progress, message is ignored");
+		err = aws_jobs_subscribe_topic_get(client, "$next", get_topic);
+		if (err) {
+			LOG_ERR("Unable to subscribe to jobs/$next/get");
+			return err;
+		}
+		break;
+	case STATE_DOWNLOADING:
+		/* Fall through */
+	case STATE_DOWNLOAD_COMPLETE:
+		if (strcmp(job_id_handling, AWS_JOB_ID_DEFAULT) == 0) {
+			return -ECANCELED;
+		}
 
-		return 0;
-	} else if (is_get_next_topic || is_get_accepted) {
-		LOG_DBG("Checking for an available job");
-		return get_job_execution(client, payload_len);
-	} else if (doc_update_accepted) {
-		return job_update_accepted(client, payload_len);
-	} else if (doc_update_rejected) {
-		LOG_ERR("Job document update was rejected");
-		return job_update_rejected(client, payload_len);
+		err = aws_jobs_subscribe_topic_update(client, job_id_handling, update_topic);
+		if (err) {
+			LOG_ERR("Error when subscribing job_id_update: %d", err);
+			return err;
+		}
+
+		LOG_DBG("Subscribed to FOTA update topic %s", log_strdup(update_topic));
+		break;
+	default:
+		break;
+	}
+
+	return 0;
+}
+
+static int on_suback_evt(struct mqtt_client *const client, uint16_t message_id)
+{
+	int err;
+
+	switch (message_id) {
+	case SUBSCRIBE_NOTIFY_NEXT:
+		LOG_DBG("Subscribed to notify-next topic");
+
+		err = aws_jobs_get_job_execution(client, "$next", get_topic);
+		if (err) {
+			LOG_ERR("aws_jobs_get_job_execution failed, error: %d", err);
+			return err;
+		}
+
+		break;
+	case SUBSCRIBE_GET:
+		LOG_DBG("subscribed to get topic");
+		break;
+	case SUBSCRIBE_JOB_ID_GET:
+		LOG_DBG("Subscribed to get next-topic");
+		break;
+	case SUBSCRIBE_JOB_ID_UPDATE:
+		LOG_DBG("subscribed to job ID update accepted/rejected topics");
+
+		enum execution_status status;
```

Comment:
Hm, does this compile without curly braces in the `case`?

Reply:
There is some switch-ception inside here but the end of this switch is further down.

I think it could be smart to do something like `} /* end switch(message_id */`  in the end on line 585

Reply:
@jtguggedal It does, at least on my end.